crates/iroha_core/src/peers_gossiper.rs

```diff
@@ -0,0 +1,151 @@
+//! Peers gossiper is actor which is responsible for gossiping addresses of peers.
+//!
+//! E.g. peer A changes address, connects to peer B,
+//! and then peer B will broadcast address of peer A to other peers.
+
+use std::{
+    collections::{HashMap, HashSet},
+    time::Duration,
+};
+
+use iroha_config::parameters::actual::TrustedPeers;
+use iroha_data_model::peer::{Peer, PeerId};
+use iroha_futures::supervisor::{Child, OnShutdown, ShutdownSignal};
+use iroha_p2p::{Broadcast, UpdatePeers};
+use iroha_primitives::{addr::SocketAddr, unique_vec::UniqueVec};
+use iroha_version::{Decode, Encode};
+use tokio::sync::mpsc;
+
+use crate::{IrohaNetwork, NetworkMessage};
+
+/// [`Gossiper`] actor handle.
+#[derive(Clone)]
+pub struct PeersGossiperHandle {
+    message_sender: mpsc::Sender<PeersGossip>,
+}
+
+impl PeersGossiperHandle {
+    /// Send [`PeersGossip`] to actor
+    pub async fn gossip(&self, gossip: PeersGossip) {
+        self.message_sender
+            .send(gossip)
+            .await
+            .expect("Gossiper must handle messages until there is at least one handle to it")
+    }
+}
+
+/// Actor which gossips peers addresses.
+pub struct PeersGossiper {
+    /// Peers provided at startup
+    initial_peers: HashMap<PeerId, SocketAddr>,
```

Comment:
can't you just put them into `gossip_peers`?

Reply:
We need both. If some malicious peer sends us invalid peer address, we will at least have initial peer address (we are trying to connect both via initial address and via gossip address).

Actually I am worried about malicious peers. Currently gossip address is just latest receiving address from any peer. Probably we need to store gossip address received from all peers (that is replace `gossip_peers: HashMap<PeerId, SocketAddr>` with `HashMap<PeerId, HashMap<PeerId, SocketAddr>>` and use majority rule

Reply:
Updated to use majority rule: https://github.com/hyperledger-iroha/iroha/pull/5176/commits/0840779d0933cdcd2303ec9cb9868c20ebac07ea